src/main/java/com/google/api/codegen/transformer/SampleFileRegistry.java

```diff
@@ -54,37 +55,53 @@ public SampleFileRegistry(SurfaceNamer namer, List<MethodSampleView> allSamples)
     }
   }
 
-  public SampleFileRegistry(SurfaceNamer namer, Collection<SampleConfig> sampleConfigs) {
+  public SampleFileRegistry(
+      SurfaceNamer namer, Map<String, List<CallingForm>> configAndMatchingForms) {
     this.namer = namer;
-    for (SampleConfig config : sampleConfigs) {
-      userProvidedIdCount.put(config.id(), userProvidedIdCount.getOrDefault(config.id(), 0) + 1);
+    for (Map.Entry<String, List<CallingForm>> entry : configAndMatchingForms.entrySet()) {
+      userProvidedIdCount.put(entry.getKey(), entry.getValue().size());
     }
   }
 
   public String getUniqueSampleId(SampleConfig config, CallingForm callingForm) {
     String userProvidedId = config.id();
-    if (userProvidedId.equals("")) {
-      userProvidedId =
+    String calculatedBaseId;
+    if (!userProvidedId.equals("")) {
+      Integer count = userProvidedIdCount.get(userProvidedId);
+      Preconditions.checkState(
+          count != null && count > 0,
+          "sample not registered. Id: %s, calling_form: %s",
+          userProvidedId,
+          callingForm);
+
+      // There is only one valid calling form for this sample config,
+      // and the user provided id becomes the final unique id
+      if (count == 1) {
+        return userProvidedId;
+      }
+
+      // The user-provided sample id contains multiple calling forms
+      // We use `{id}_{calling_form}` as the new base id
+      calculatedBaseId =
+          Name.anyLower(userProvidedId, callingForm.toLowerUnderscore()).toLowerUnderscore();
+    } else {
+
+      // User did not provide a sample id. Use `sample_{method_name}`
+      // as a base id.
+      calculatedBaseId =
           "sample_"
               + Name.upperCamel(config.methodConfig().getMethodModel().getSimpleName())
                   .toLowerUnderscore();
     }
-    Integer count = userProvidedIdCount.get(userProvidedId);
-    Preconditions.checkState(count != null && count > 0, "Sample not registered.");
 
-    if (count == 1) {
-      return userProvidedId;
+    // Find a unique id based on base id and save the unique id.
+    String calculatedId = calculatedBaseId;
+    while (userProvidedIdCount.containsKey(calculatedId)) {
+      int suffix = firstNonNull(usedSuffixes.get(calculatedBaseId), 0) + 1;
+      usedSuffixes.put(calculatedId, suffix);
+      calculatedId = calculatedBaseId + suffix;
```

Comment:
should we have an underscore before the numeric suffix, so it reads better in snake case?

Reply:
Sure.